### PR TITLE
Creating vector-type calculators and adding median calculator

### LINKED ISSUE #17799: Add Median Calculator
## Reason

The median of stochastic results is useful.

## Design

Create `Median` object that inherits from `Calculator` that correctly operates with a distributed vector.

## Impact

New capability.

### PR DESCRIPTION
**Pull request by zachmprince**
This PR does four things:
1. Improve computation of bootstrap statistics by reducing parallel communication
1. Generalizing calculators a bit by handling vector-type input-output data
1. Creating a median calculator for distributed data
1. Making SobolCalculator a derived class of Calculator and getting it work with distributed data.

Details on how each one functions:

1. Bootstrapping is performed by randomly selecting indices of a vector of data (resampling) and recomputing the statistics. Previously, this was done by generating global indices on each processor, communicating the required indices to the appropriate process, and gathering the data back to the process requesting it. This essentially performed two pretty expensive communications: sending the indices and receiving the data back. I realized the second communication wasn't necessary since calculators don't care about the order of the data and most calculators can act on on the data in-place. Therefore, I reworked the calculators to be able to compute within a loop of data that looks like this:
```cpp
auto calculator = makeCalculator<InType, OutType>(*this, type);
calculator->initializeCalculator();
for (const typename InType::value_type & val : data)
  calculator->updateCalculator(data); 
calculator->finalizeCalculator(is_distributed);
OutType stat = calculator->getValue();
```
Although there is a utility in the base class that does this automatically:
```cpp
auto calculator = makeCalculator<InType, OutType>(*this, type);
OutType stat = calculator->compute(data, is_distributed);
```
This might be a little detrimental for calculators that need the entire set of data (namely Median) which will basically create a copy of the vector. But we were doing this anyway with previously resampling techniques.

2. Creation of a `VectorCalculator<std::vector<InType>, std::vector<OutType>, CalcType>` which creates a `CalcType<InType, OutType>` for each each element of input data vector. 

3. Median calculators typically work my sorting the data and choosing the middle value. However, this really only works on serial vectors since there would need to be an aggressive amount of communication to sort a parallel vector. Instead, I came up with an algorithm that works by choosing a value, seeing how many values are greater and less than that value, removing the whichever is smaller, and repeating this process until the correct number of values are less than and greater than. The iterative process requires a broadcast and reduction each iteration. I believe this is tractable with suitably random data with lots of values on each process, although I have not looked at the performance. Still better than gather and sort IMO.

4. Previously, SobolCalculator errored when trying to act on distributed data. I tried doing something fancy to keep the data distributed, but this was very difficult since it bascially requires a product of two parallel dense matrices. I feel figuring this out is the best way, but in the mean time, I just did a gather and broadcast. This is still better than required replicated data as was required before. Also, deriving SobolCalculator from Calculator allows bootstrapping. However, I tried testing this and could not get reasonable results. There must be something I'm not understanding about the Sobol algorithm, so I'll leave this to future work once I read more about bootstrapping.

In conclusion, everything seems to work but I'm not totally convinced its the correct design. I do want to make a tutorial on how to make a new calculator with a new data type, but I want to get feedback on the design first. Thank you, and have a nice day!

Closes #17799 #17390

### CODE REVIEW

File: framework/include/utils/Shuffle.h [hunk 1]
```diff
@@ -385,6 +419,73 @@ MooseUtils::resample(const std::vector<T> & data,
   return replicate;
 }
 
+template <typename T, typename ActionFunctor>
+void
+MooseUtils::resampleAct(const std::vector<T> & data,
+                        const ActionFunctor & act,
+                        MooseRandom & generator,
+                        const std::size_t seed_index,
+                        const libMesh::Parallel::Communicator * comm_ptr)
```
Comment (zachmprince) [inline framework/include/utils/Shuffle.h:428]: @loganharbour This is the only function that you need to worry about.

Comment (loganharbour) [reply framework/include/utils/Shuffle.h:428]: What a beaut.

File: modules/stochastic_tools/include/utils/BootstrapCalculators.h [hunk 2]
```diff
@@ -36,22 +48,21 @@ MooseEnum makeBootstrapCalculatorEnum();
...
+class BootstrapCalculator : public libMesh::ParallelObject
 {
 public:
   BootstrapCalculator(const libMesh::ParallelObject & other,
                       const std::string & name,
                       const std::vector<Real> & levels,
                       unsigned int replicates,
                       unsigned int seed,
-                      const StochasticTools::Calculator<InType, OutType> & calc);
+                      StochasticTools::Calculator<InType, OutType> & calc);
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/BootstrapCalculators.h:59]: This makes me nervous, computing the Bootstrap statistics should not alter the state of the Calculator. Why is this needed?

File: modules/stochastic_tools/include/utils/Calculators.h [hunk 3]
```diff
@@ -125,7 +212,525 @@ class L2Norm : public Calculator<InType, OutType>
...
+
+  /**
+   * These are functions that do modifications on the value that do not depend
+   * on the in-type @tparam T1.
+   */
+  ///@{
+  /// Set the value to zero
+  void zero() { _value = T2(); }
+  /// Divide the value by a interger value
+  void divide(dof_id_type num) { _value /= num; }
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/Calculators.h:245]: Why is this an integer only?

File: modules/stochastic_tools/include/utils/Calculators.h [hunk 4]
```diff
@@ -125,7 +212,525 @@ class L2Norm : public Calculator<InType, OutType>
...
+  void max(const T1 & a) { _value = std::max(static_cast<T2>(a), _value); }
+  ///@}
+
+  /**
+   * These are overloaded operators that modify the value based on out-type values
+   */
+  ///@{
+  CalculatorValue<T1, T2> & operator+=(const T2 & b)
+  {
+    _value += b;
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/Calculators.h:276]: Could you utilize the above methods, so that these do not need to be defined in the specializations and work for all instantiated CalculatorValues?

File: modules/stochastic_tools/include/utils/Calculators.h [hunk 5]
```diff
@@ -125,7 +212,525 @@ class L2Norm : public Calculator<InType, OutType>
...
+  {
+    _value -= b;
+    return *this;
+  }
+  CalculatorValue<T1, T2> & operator/=(const T2 & b)
+  {
+    _value /= b;
+    return *this;
+  }
+  bool less_than(const T2 & b) const { return _value < b; }
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/Calculators.h:289]: Move this up wit the named functions to keep them together. This should also be `lessThan`

File: modules/stochastic_tools/include/utils/Calculators.h (hunk 4 above)
Comment (aeslaughter) [reply modules/stochastic_tools/include/utils/Calculators.h:276]: Also, what about `greaterThan` and `equal`?

File: modules/stochastic_tools/include/utils/Calculators.h [hunk 6]
```diff
@@ -125,7 +212,525 @@ class L2Norm : public Calculator<InType, OutType>
...
+  ///@}
+
+  /**
+   * These are MPI operations
+   */
+  ///@{
+  void sum(const libMesh::Parallel::Communicator & comm) { comm.sum(_value); }
+  void min(const libMesh::Parallel::Communicator & comm) { comm.min(_value); }
+  void max(const libMesh::Parallel::Communicator & comm) { comm.max(_value); }
+  void broadcast(const libMesh::Parallel::Communicator & comm, processor_id_type root_id)
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/Calculators.h:299]: Should these be given a prefix so they don't get confused with the other function?

`parallelMin`, `mpiMin`, `minMPI`, `minParallel`, `minComm` ???

File: framework/include/utils/Shuffle.h [hunk 7]
```diff
@@ -385,6 +419,73 @@ MooseUtils::resample(const std::vector<T> & data,
   return replicate;
 }
 
+template <typename T, typename ActionFunctor>
+void
+MooseUtils::resampleAct(const std::vector<T> & data,
```
Comment (aeslaughter) [inline framework/include/utils/Shuffle.h:424]: I like the concept, but don't like the name. It doesn't help me understand what it does, but I can't think of something better. `resampleWithFunctor`?

File: modules/stochastic_tools/include/utils/BootstrapCalculators.h [hunk 8]
```diff
@@ -74,30 +88,219 @@ class Percentile : public BootstrapCalculator<InType, OutType>
...
+class BiasCorrectedAccelerated<InType, Real> : public BootstrapCalculator<InType, Real>
+{
+public:
+  using BootstrapCalculator<InType, Real>::BootstrapCalculator;
+  virtual std::vector<Real> compute(const InType &, const bool) override;
 
 private:
   // Compute the acceleration, see Efron and Tibshirani (2003), Ch. 14, Eq. 14.15, p 186.
-  Real acceleration(const InType &, const bool) const;
+  Real acceleration(const InType &, const bool);
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/BootstrapCalculators.h:124]: Why is `const` being removed. Computing things for one InType should not change the state of the calculator? You should be able to call these as much as you want without any worry something is altering the calculation.

Comment (aeslaughter) [reply modules/stochastic_tools/include/utils/BootstrapCalculators.h:124]: I am seeing why this is done now, the Calculator::initialize/finalize. Is there some way we can assert the calculator state before resampleAct is called?

Comment (zachmprince) [reply modules/stochastic_tools/include/utils/BootstrapCalculators.h:124]: I think I have an idea. I can make enum like this:
```cpp
enum StochasticTools:CalculatorState
{
  NONE;
  INITIALIZED;
  FINALIZED;
};
```
And have member variable in `Calculator`:
```cpp
  CalculatorState _state = CalculatorState::NONE;
```

And it will error if:
- `update` or `finalize` is called and `_state != CalculatorState::INITIALIZED`
- `get` is called and `_state != CalculatorState::FINALIZED`

In order to make the calculator `const` all the member variable would have to be `mutable`.  This might be okay with this state check.

File: framework/include/utils/Shuffle.h (hunk 7 above)
Comment (zachmprince) [reply framework/include/utils/Shuffle.h:424]: Haha, that was my original name, but I thought it was too long. I will change back.

File: modules/stochastic_tools/include/utils/Calculators.h (hunk 4 above)
Comment (zachmprince) [reply modules/stochastic_tools/include/utils/Calculators.h:276]: I added functions based on what was needed by the calculators. `a.greaterThan(b)` is the same as `b.lessThan(a)` and `a.equal(b)` is the same as `!(a.lessThan(b) || b.lessThan(a))`.

I'm starting to think there are too many subtleties in this implementation. You basically have to understand how the calculator is using these functions to define them correctly, which is what I was trying to avoid in the first place. I'm thinking of going back and doing specializations of the calculators themselves.

File: modules/stochastic_tools/include/utils/Calculators.h [hunk 9]
```diff
@@ -43,80 +49,166 @@ class Calculator : public libMesh::ParallelObject
...
+  void update(const typename InType::value_type &);
+  void finalize(bool);
+  OutType get() const;
+
   const std::string & name() const { return _name; }
 
+protected:
+  virtual void initializeCalculator() = 0;
+  virtual void updateCalculator(const typename InType::value_type &) = 0;
+  virtual void finalizeCalculator(bool) = 0;
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/Calculators.h:69]: I think these functions should be flipped, the internal methods should have the Calculator prefix (or name them intializeInternal). The non-prefixed is what should be overridden, mainly because that is how we do it with other objects in MOOSE.

**Review by aeslaughter — CHANGES_REQUESTED** [review]

File: modules/stochastic_tools/include/utils/Calculators.h (hunk 9 above)
Comment (zachmprince) [reply modules/stochastic_tools/include/utils/Calculators.h:69]: Sorry, not sure what you mean here. The public functions do not have the suffix, BoostrapCalculator calls `initialize`, `update`, and `finalize`. The internal protected functions have the suffix and are the ones meant to be overridden. I'm not sure how to write it such that the public functions are the ones being overridden, unless every calculator has the "state" checks.

Comment (aeslaughter) [reply modules/stochastic_tools/include/utils/Calculators.h:69]: I was trying to say the "initialize", "update", and "finalize" should be the ones meant to be overridden, to be consistent with other objects in MOOSE. Have BootstrapCalculator call the methods with the prefix instead.

Comment (zachmprince) [reply modules/stochastic_tools/include/utils/Calculators.h:69]: Ohh, got it. I will update right now.

File: modules/stochastic_tools/include/utils/Calculators.h [hunk 10]
```diff
@@ -43,80 +49,156 @@ class Calculator : public libMesh::ParallelObject
...
   {
   }
 
   virtual ~Calculator() = default;
-  virtual OutType compute(const InType &, bool) const = 0;
+  OutType compute(const InType &, bool);
+
+  void initializeCalculator();
+  void updateCalculator(const typename InType::value_type &);
+  void finalizeCalculator(bool);
```
Comment (aeslaughter) [inline modules/stochastic_tools/include/utils/Calculators.h:61]: Can you add some documentation on these methods as well as on the pure virtual methods as well.

**Review by aeslaughter — CHANGES_REQUESTED** [review]
One more small change.

File: modules/stochastic_tools/include/utils/Calculators.h (hunk 10 above)
Comment (zachmprince) [reply modules/stochastic_tools/include/utils/Calculators.h]: Oops, yeah, definitely need to do that.

**Review by aeslaughter — APPROVED** [review]
